### PR TITLE
feat: support pytorch-optimizer training optimizers

### LINKED ISSUE #89: Support arbitrary optimizer from pytorch-optimizer
pytorch-optimizer implements 100+ optimizers: https://pytorch-optimizers.readthedocs.io/en/latest/

Would be great to have native support to all.

### PR DESCRIPTION
**Pull request by mfazrinizar**
## What does this PR do?

This PR adds configurable optimizer support to RF-DETR training while keeping the existing AdamW behavior as the default. `optimizer="adamw"` continues to use RF-DETR's built-in fused `torch.optim.AdamW` path, and non-default optimizer names are resolved through `pytorch-optimizer`, for example `optimizer="lion"` or `optimizer="pytorch_optimizer:adamw"`.

The implementation preserves RF-DETR parameter groups and layer-wise learning rates by building optimizers from the existing `get_param_dict()` output. It also adds `optimizer_kwargs` so users can pass optimizer-specific arguments such as AdamW `betas` or Lion `weight_decouple` without overriding RF-DETR-managed values like `params`, `lr`, `weight_decay`, or `fused`.

It also wires the options through the public training API, keeps PTL-only optimizer config out of the legacy namespace, adds focused tests, documents the new parameters, and includes `pytorch-optimizer` in the training extra.

**Related Issue(s):** Closes #89

## Type of Change

- New feature (non-breaking change that adds functionality)

## Testing

- [x] I have tested this change locally
- [x] I have added/updated tests for this change

**Test details:**

Added and updated tests cover:

- `TrainConfig` defaults, accepted values, empty-name rejection, and reserved `optimizer_kwargs` rejection
- default AdamW behavior remaining backward compatible
- `optimizer_kwargs` forwarding to RF-DETR's default AdamW optimizer
- custom optimizer loading through `pytorch-optimizer`
- RF-DETR parameter groups and layer-wise learning rates being preserved for custom optimizers
- custom optimizer kwargs forwarding
- explicit `pytorch_optimizer:` prefix support, including opting into the external AdamW implementation
- missing dependency and invalid optimizer-name error handling
- real `pytorch-optimizer` Lion construction smoke test
- `RFDETR.train(optimizer=..., optimizer_kwargs=...)` forwarding into `get_train_config()`
- optimizer-only config staying out of the legacy namespace

Local validation was run with `PYTHONPATH=src` in the `rfdetr` conda environment via the environment Python executable.

Real optimizer smoke tests passed for:

- direct `pytorch-optimizer` Lion construction and one optimizer step with a PyTorch parameter group
- RF-DETR's `_build_pytorch_optimizer()` constructing Lion with RF-DETR-style parameter groups and running one optimizer step with `optimizer_kwargs={"weight_decouple": True}`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors
- [x] I have updated the documentation accordingly (if applicable)

## Additional Context

The implementation intentionally uses `pytorch_optimizer.load_optimizer()` instead of `create_optimizer()` so RF-DETR keeps its existing parameter grouping, backbone learning rates, and scheduler behavior. Some specialized optimizers may still require optimizer-specific kwargs; initialization errors include a hint to check the selected optimizer's supported arguments.

### CODE REVIEW

**Comment by Alarmod** [comment]
I tested Yolo26 with AdamW and MuSGD 
https://github.com/ultralytics/ultralytics/issues/23789

```
from ultralytics.optim.muon import MuSGD
# Initialize optimizer with specific parameter groups
# Use 'use_muon=True' only for 2D+ tensors for the hybrid effect
optimizer = MuSGD(model.parameters(), lr=0.01, momentum=0.9)
```
It will be compatible with external optimizer?

**Comment by mfazrinizar** [comment]
@Alarmod it's compatible with external optimizers provided by pytorch-optimizer that accept normal PyTorch param groups. However, it's not automatically compatible with arbitrary imported optimizer classes like Ultralytics MuSGD yet. That could be a follow-up with a small adapter/registry and dedicated tests, interesting to support it. Already working on it.

**Comment by isaacrob** [comment]
let's please not link to ultralytics due to licensing issues. do think supporting pytorch optimizers makes sense! :) would love to see some muon results .. little tricky tho cause you have to make sure that muon is operating on only the appropriate layers

**Comment by Borda** [comment]
FYI, there is this package with many optimizers (unfortunately not maintained for a while) - https://github.com/jettify/pytorch-optimizer
And here is an overview - https://github.com/kozistr/pytorch_optimizer

My take is to pass the optimizer as a partially initialized object, so we would supply in the module only params...

**Comment by Alarmod** [comment]
> let's please not link to ultralytics due to licensing issues. do think supporting pytorch optimizers makes sense! :) would love to see some muon results .. little tricky tho cause you have to make sure that muon is operating on only the appropriate layers

I am currently writing a scientific article about RF-DETR and Yolo26 with AdamW and Muon. Search for brain tumors...

**Comment by Alarmod** [comment]
@mfazrinizar @isaacrob @Borda 

Command to install the fork:
pip3 install git+https://github.com/mfazrinizar/rf-detr.git@feat/pytorch-optimizer-train

In the end, I tested a large number of optimizers from pytorch-optimizer in my task with MRI data. Here is a list of the best ones:
AdamW -- mAP50-95 -- 0.594, weight_decay=1e-2
**AdamW -- mAP50-95 -- 0.597, weight_decay=1e-4 (default in RF-DETR)**
RAdam -- mAP50-95 -- 0.600
AdaBelief -- mAP50-95 -- 0.601
**AdamP -- mAP50-95 -- 0.604, weight_decay=1e-2**
AdamP -- mAP50-95 -- 0.602, weight_decay=1e-4 (default, parameter taken from the class constructor in the RF-DETR code)

So, I managed to increase mAP50-95 from 0.597 to 0.604, which is a 0.7% absolute improvement. This is not a huge gain, but still an achievement.

**Comment by Borda** [comment]
Yes, we will land it, just need to polish the best UX pattern...

**Comment by mfazrinizar** [comment]
@Borda just asking the state of this PR's progress. Regarding the the public API, could you confirm the intended optimizer UX? You suggested passing a partially initialized optimizer and having RF-DETR supply only `params`.

Should the factory receive RF-DETR’s existing layer-wise parameter groups or raw `model.parameters()`? And should `lr` / `weight_decay` live in the partial? I’d keep callable factories Python-API-only initially because they are not naturally config/CLI serializable. Let me know when you have the directions.

**Comment by Borda** [comment]
It is planed for next release...
